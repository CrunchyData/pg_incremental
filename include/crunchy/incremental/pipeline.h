#pragma once

#define SEQUENCE_RANGE_PIPELINE 's'
#define TIME_INTERVAL_PIPELINE 't'
#define FILE_LIST_PIPELINE 'f'

typedef char PipelineType;

/*
 * PipelineDesc describes a pipeline.
 */
typedef struct PipelineDesc
{
	/* name of the pipeline */
	char	   *pipelineName;

	/* type of the pipeline */
	PipelineType pipelineType;

	/* user ID of the pipeline owner */
	Oid			ownerId;

	/* OID of the source relation or sequence */
	Oid			sourceRelationId;

	/* command to run for the pipeline */
	char	   *command;

	/* search_path to use during execution */
	char	   *searchPath;
}			PipelineDesc;

PipelineDesc *ReadPipelineDesc(char *pipelineName);
