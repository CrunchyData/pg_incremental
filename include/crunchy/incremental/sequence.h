#pragma once

void		InitializeSequencePipelineState(char *pipelineName, Oid sequenceId);
void		UpdateLastProcessedSequenceNumber(char *pipelineName, int64 lastSequenceNumber);
void		ExecuteSequenceRangePipeline(char *pipelineName, char *command);
Oid			FindSequenceForRelation(Oid relationId);
