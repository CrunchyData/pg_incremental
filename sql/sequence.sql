create extension pg_incremental cascade;
create schema sequence;
set search_path to sequence;

-- create a source table
create table events (
  event_id bigint generated always as identity,
  event_time timestamptz default now(),
  client_id bigint,
  path text,
  response_time double precision
);

-- BRIN indexes are highly effective in selecting new ranges
create index on events using brin (event_id);

-- generate some random inserts
insert into events (client_id, path, response_time)
select s, '/page-' || (s % 3), random() from generate_series(1,100) s;

-- create a summary table to pre-aggregate the number of events per day
create table events_agg (
  day timestamptz,
  event_count bigint,
  primary key (day)
);

select incremental.create_sequence_pipeline('event-aggregation',
  source_table_name := 'events',
  schedule := NULL,
  command := $$
  insert into events_agg
  select date_trunc('day', event_time), count(*)
  from events
  where event_id between $1 and $2
  group by 1
  on conflict (day) do update set event_count = events_agg.event_count + excluded.event_count;
  $$);

select count(*) from events;
select sum(event_count) from events_agg;

call incremental.execute_pipeline('event-aggregation');

insert into events (client_id, path, response_time)
select s, '/page-' || (s % 3), random() from generate_series(1,100) s;

call incremental.execute_pipeline('event-aggregation');

select count(*) from events;
select sum(event_count) from events_agg;

create table events_json (id bigint generated by default as identity, payload jsonb);

/* periodically unpack the new JSON objects into the events table */
select incremental.create_sequence_pipeline('unpack-json', 'events_json', 
  schedule := NULL,
  command := $$
    insert into events (client_id, event_time, response_time)
    select '3', (payload->>'created_at')::timestamptz, random()
    from events_json
    where id between $1 and $2
  $$
);

insert into events_json (payload) values ('{"created_at":"2024-01-01 03:00:00"}');

call incremental.execute_pipeline('unpack-json');
call incremental.execute_pipeline('event-aggregation');

select count(*) from events;
select sum(event_count) from events_agg;

drop schema sequence cascade;
drop extension pg_incremental;
