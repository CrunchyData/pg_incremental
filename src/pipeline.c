#include "postgres.h"
#include "fmgr.h"
#include "miscadmin.h"

#include "catalog/dependency.h"
#include "catalog/pg_authid.h"
#include "crunchy/incremental/cron.h"
#include "crunchy/incremental/pipeline.h"
#include "crunchy/incremental/sequence.h"
#include "executor/spi.h"
#include "tcop/tcopprot.h"
#include "tcop/utility.h"
#include "utils/builtins.h"
#include "utils/lsyscache.h"
#include "utils/ruleutils.h"

static void CreatePipeline(char *pipelineName, PipelineType pipelineType, Oid sequenceId,
						   char *command, char *schedule, bool executeImmediately);
static void InsertPipeline(char *pipelineName, PipelineType pipelineType, Oid sourceRelationId,
						   char *command);
static void EnsurePipelineOwner(char *pipelineName, Oid ownerId);
static void ExecutePipeline(char *pipelineName, PipelineType pipelineType, char *command);
static void ResetPipeline(char *pipelineName, PipelineType pipelineType);
static void DeletePipeline(char *pipelineName);
static char *GetCronJobNameForPipeline(char *pipelineName);
static char *GetCronCommandForPipeline(char *pipelineName);

static Query *ParseQuery(char *command, List *paramTypes);
static char *DeparseQuery(Query *query);
static void ExecuteCommand(char *commandString);


PG_FUNCTION_INFO_V1(incremental_create_sequence_pipeline);
PG_FUNCTION_INFO_V1(incremental_execute_pipeline);
PG_FUNCTION_INFO_V1(incremental_reset_pipeline);
PG_FUNCTION_INFO_V1(incremental_drop_pipeline);


/*
 * incremental_create_sequence_pipeline creates a new pipeline that tracks
 * a sequence.
 */
Datum
incremental_create_sequence_pipeline(PG_FUNCTION_ARGS)
{
	/*
	 * create_sequence_pipeline is not strict because the last argument can be
	 * NULL, so check the arguments that cannot be NULL.
	 */
	if (PG_ARGISNULL(0))
		ereport(ERROR, (errmsg("pipeline_name cannot be NULL")));
	if (PG_ARGISNULL(1))
		ereport(ERROR, (errmsg("sequence_name cannot be NULL")));
	if (PG_ARGISNULL(2))
		ereport(ERROR, (errmsg("command cannot be NULL")));

	char	   *pipelineName = text_to_cstring(PG_GETARG_TEXT_P(0));
	Oid			sequenceId = PG_GETARG_OID(1);
	char	   *command = text_to_cstring(PG_GETARG_TEXT_P(2));
	char	   *schedule = PG_ARGISNULL(3) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(3));

	switch (get_rel_relkind(sequenceId))
	{
		case RELKIND_SEQUENCE:
			{
				Oid			relationId = InvalidOid;
				int32		columnNumber = 0;

				if (!sequenceIsOwned(sequenceId, DEPENDENCY_AUTO, &relationId, &columnNumber))
				{
					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
									errmsg("only sequences that are owned by a table are supported")));
				}

				break;
			}

		case RELKIND_RELATION:
		case RELKIND_FOREIGN_TABLE:
		case RELKIND_PARTITIONED_TABLE:
			{
				/* user entered a table name, see if it has a single sequence */
				sequenceId = FindSequenceForRelation(sequenceId);
				break;
			}

		default:
			{
				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
								errmsg("%s is not a table or sequence", get_rel_name(sequenceId))));
			}
	}

	CreatePipeline(pipelineName, SEQUENCE_PIPELINE, sequenceId, command, schedule, true);

	PG_RETURN_VOID();
}


/*
 * incremental_execute_pipeline executes a pipeline to its initial state.
 */
Datum
incremental_execute_pipeline(PG_FUNCTION_ARGS)
{
	char	   *pipelineName = text_to_cstring(PG_GETARG_TEXT_P(0));
	PipelineDesc *pipelineDesc = ReadPipelineDesc(pipelineName);

	EnsurePipelineOwner(pipelineName, pipelineDesc->ownerId);
	ExecutePipeline(pipelineName, pipelineDesc->pipelineType, pipelineDesc->command);

	PG_RETURN_VOID();
}


/*
 * incremental_reset_pipeline reset a pipeline to its initial state.
 */
Datum
incremental_reset_pipeline(PG_FUNCTION_ARGS)
{
	char	   *pipelineName = text_to_cstring(PG_GETARG_TEXT_P(0));
	PipelineDesc *pipelineDesc = ReadPipelineDesc(pipelineName);

	EnsurePipelineOwner(pipelineName, pipelineDesc->ownerId);
	ResetPipeline(pipelineName, pipelineDesc->pipelineType);

	PG_RETURN_VOID();
}


/*
 * incremental_drop_pipeline drops a pipeline by name.
 */
Datum
incremental_drop_pipeline(PG_FUNCTION_ARGS)
{
	char	   *pipelineName = text_to_cstring(PG_GETARG_TEXT_P(0));
	PipelineDesc *pipelineDesc = ReadPipelineDesc(pipelineName);

	EnsurePipelineOwner(pipelineName, pipelineDesc->ownerId);
	DeletePipeline(pipelineName);

	UnscheduleCronJob(GetCronJobNameForPipeline(pipelineName));

	PG_RETURN_VOID();
}


/*
 * CreatePipeline sets up a pipeline for a table with a sequence.
 */
static void
CreatePipeline(char *pipelineName, PipelineType pipelineType, Oid sourceRelationId,
			   char *command, char *schedule, bool executeImmediately)
{
	/* determine the parameter list for the query */
	List	   *paramTypes = NIL;

	switch (pipelineType)
	{
		case SEQUENCE_PIPELINE:
			paramTypes = list_make2_oid(INT8OID, INT8OID);
			break;

		case FILE_PIPELINE:
			paramTypes = list_make1_oid(TEXTARRAYOID);
			break;

		default:
			elog(ERROR, "unknown pipeline type: %c", pipelineType);
	}

	/* sanitize the query */
	Query	   *parsedQuery = ParseQuery(command, paramTypes);
	char	   *sanitizedCommand = DeparseQuery(parsedQuery);

	InsertPipeline(pipelineName, SEQUENCE_PIPELINE, sourceRelationId, sanitizedCommand);

	switch (pipelineType)
	{
		case SEQUENCE_PIPELINE:
			InitializeSequencePipelineState(pipelineName, sourceRelationId);
			break;

		case FILE_PIPELINE:
			/* initial state is empty */
			break;

		default:
			elog(ERROR, "unknown pipeline type: %c", pipelineType);
	}

	if (executeImmediately)
	{
		ExecutePipeline(pipelineName, pipelineType, sanitizedCommand);
	}

	if (schedule != NULL)
	{
		char	   *jobName = GetCronJobNameForPipeline(pipelineName);
		char	   *cronCommand = GetCronCommandForPipeline(pipelineName);

		int64		jobId = ScheduleCronJob(jobName, schedule, cronCommand);

		ereport(NOTICE, (errmsg("pipeline %s: scheduled cron job  with ID " INT64_FORMAT
								" and schedule %s",
								pipelineName, jobId, schedule)));
	}
}


/*
 * InsertPipeline adds a new pipeline.
 */
static void
InsertPipeline(char *pipelineName, PipelineType pipelineType, Oid sourceRelationId,
			   char *command)
{
	Oid			savedUserId = InvalidOid;
	int			savedSecurityContext = 0;

	/*
	 * Switch to superuser in case the current user does not have write
	 * privileges for the pipelines table.
	 */
	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
	SetUserIdAndSecContext(BOOTSTRAP_SUPERUSERID, SECURITY_LOCAL_USERID_CHANGE);

	char	   *query =
		"insert into incremental.pipelines "
		"(pipeline_name, pipeline_type, owner_id, source_relation, command) "
		"values ($1, $2, $3, $4, $5)";

	bool		readOnly = false;
	int			tupleCount = 0;
	int			argCount = 5;
	Oid			argTypes[] = {TEXTOID, CHAROID, OIDOID, OIDOID, TEXTOID};
	Datum		argValues[] = {
		CStringGetTextDatum(pipelineName),
		CharGetDatum(pipelineType),
		ObjectIdGetDatum(savedUserId),
		ObjectIdGetDatum(sourceRelationId),
		CStringGetTextDatum(command)
	};
	char	   *argNulls = "     ";

	SPI_connect();
	SPI_execute_with_args(query,
						  argCount,
						  argTypes,
						  argValues,
						  argNulls,
						  readOnly,
						  tupleCount);
	SPI_finish();

	SetUserIdAndSecContext(savedUserId, savedSecurityContext);
}


/*
 * ReadPipelineCommand returns a full description of a pipeline.
 */
PipelineDesc *
ReadPipelineDesc(char *pipelineName)
{
	PipelineDesc *pipelineDesc = (PipelineDesc *) palloc0(sizeof(PipelineDesc));

	Oid			savedUserId = InvalidOid;
	int			savedSecurityContext = 0;

	/*
	 * Switch to superuser in case the current user does not have read
	 * privileges for the pipelines table.
	 */
	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
	SetUserIdAndSecContext(BOOTSTRAP_SUPERUSERID, SECURITY_LOCAL_USERID_CHANGE);

	MemoryContext callerContext = CurrentMemoryContext;

	char	   *query =
		"select pipeline_type, owner_id, source_relation, command "
		"from incremental.pipelines where pipeline_name = $1";

	bool		readOnly = false;
	int			tupleCount = 0;
	int			argCount = 1;
	Oid			argTypes[] = {TEXTOID};
	Datum		argValues[] = {
		CStringGetTextDatum(pipelineName)
	};
	char	   *argNulls = " ";

	/* we do not switch user, because we want to preserve permissions */

	SPI_connect();
	SPI_execute_with_args(query,
						  argCount,
						  argTypes,
						  argValues,
						  argNulls,
						  readOnly,
						  tupleCount);

	if (SPI_processed <= 0)
		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT),
						errmsg("no such pipeline named \"%s\"", pipelineName)));

	TupleDesc	rowDesc = SPI_tuptable->tupdesc;
	HeapTuple	row = SPI_tuptable->vals[0];

	bool		isNull = false;
	Datum		pipelineTypeDatum = SPI_getbinval(row, rowDesc, 1, &isNull);
	Datum		ownerIdDatum = SPI_getbinval(row, rowDesc, 2, &isNull);
	Datum		sourceRelationDatum = SPI_getbinval(row, rowDesc, 3, &isNull);
	Datum		commandDatum = SPI_getbinval(row, rowDesc, 4, &isNull);

	MemoryContext spiContext = MemoryContextSwitchTo(callerContext);

	pipelineDesc->pipelineName = pstrdup(pipelineName);
	pipelineDesc->pipelineType = DatumGetChar(pipelineTypeDatum);
	pipelineDesc->ownerId = DatumGetObjectId(ownerIdDatum);
	pipelineDesc->sourceRelationId = DatumGetObjectId(sourceRelationDatum);
	pipelineDesc->command = TextDatumGetCString(commandDatum);

	MemoryContextSwitchTo(spiContext);

	SPI_finish();

	return pipelineDesc;
}



/*
 * EnsurePipelineOwner throws an error if the current user is not
 * superuser and not the pipeline owner.
 */
static void
EnsurePipelineOwner(char *pipelineName, Oid ownerId)
{
	if (superuser())
		return;

	if (ownerId != GetUserId())
		ereport(ERROR, (errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
						errmsg("permission denied for pipeline %s", pipelineName)));
}


/*
 * ExecutePipeline executes a pipeline.
 */
static void
ExecutePipeline(char *pipelineName, PipelineType pipelineType, char *command)
{
	char	   *fullCommand = NULL;

	switch (pipelineType)
	{
		case SEQUENCE_PIPELINE:
			fullCommand = GetSequencePipelineCommand(pipelineName, command);
			break;

		default:
			elog(ERROR, "unknown pipeline type: %c", pipelineType);
	}

	ExecuteCommand(fullCommand);
}


/*
 * ResetPipeline reset a pipeline to its initial state.
 */
static void
ResetPipeline(char *pipelineName, PipelineType pipelineType)
{
	switch (pipelineType)
	{
		case SEQUENCE_PIPELINE:
			UpdateLastProcessedSequenceNumber(pipelineName, 0);
			break;

		default:
			elog(ERROR, "unknown pipeline type: %c", pipelineType);
	}
}


/*
 * DeletePipeline removes a pipeline from the pipeline.pipelines table.
 */
static void
DeletePipeline(char *pipelineName)
{
	Oid			savedUserId = InvalidOid;
	int			savedSecurityContext = 0;

	/*
	 * Switch to superuser in case the current user does not have write
	 * privileges for the pipelines table.
	 */
	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
	SetUserIdAndSecContext(BOOTSTRAP_SUPERUSERID, SECURITY_LOCAL_USERID_CHANGE);

	char	   *query =
		"delete from incremental.pipelines "
		"where pipeline_name operator(pg_catalog.=) $1";

	bool		readOnly = false;
	int			tupleCount = 0;
	int			argCount = 1;
	Oid			argTypes[] = {TEXTOID};
	Datum		argValues[] = {
		CStringGetTextDatum(pipelineName)
	};
	char	   *argNulls = " ";

	SPI_connect();
	SPI_execute_with_args(query,
						  argCount,
						  argTypes,
						  argValues,
						  argNulls,
						  readOnly,
						  tupleCount);

	SPI_finish();

	SetUserIdAndSecContext(savedUserId, savedSecurityContext);
}


/*
 * GetCronJobNameForPipeline returns the name of the cron job to use for a given pipeline.
 */
static char *
GetCronJobNameForPipeline(char *pipelineName)
{
	return psprintf("pipeline:%s", pipelineName);
}


/*
 * GetCronCommandForPipeline returns the command of the cron job to use for a given pipeline.
 */
static char *
GetCronCommandForPipeline(char *pipelineName)
{
	return psprintf("select incremental.execute_pipeline(%s)",
					quote_literal_cstr(pipelineName));
}


/*
 * ParseQuery parses a query string.
 *
 * The function parses the query and returns the query tree.
 */
static Query *
ParseQuery(char *command, List *paramTypes)
{
	List	   *parseTreeList = pg_parse_query(command);

	if (list_length(parseTreeList) != 1)
		ereport(ERRCODE_FEATURE_NOT_SUPPORTED,
				(errmsg("pg_pipeline can only execute a single query in a pipeline")));

	Oid		   *params = (Oid *) palloc(list_length(paramTypes) * sizeof(Oid));
	ListCell   *paramTypeCell = NULL;
	int			paramCount = 0;

	foreach(paramTypeCell, paramTypes)
	{
		params[paramCount++] = lfirst_oid(paramTypeCell);
	}

	RawStmt    *rawStmt = (RawStmt *) linitial(parseTreeList);
	List	   *queryTreeList =
		pg_analyze_and_rewrite_fixedparams(rawStmt, command, params, paramCount, NULL);

	/* we already checked parseTreeList lenght above */
	Assert(list_length(queryTreeList) == 1);

	return (Query *) linitial(queryTreeList);
}


/*
 * DeparsQuery deparses a Query AST.
 */
static char *
DeparseQuery(Query *query)
{
	int			save_nestlevel = NewGUCNestLevel();

	(void) set_config_option("search_path", "pg_catalog",
							 PGC_USERSET, PGC_S_SESSION,
							 GUC_ACTION_SAVE, true, 0, false);

	bool		pretty = false;
	char	   *newQuery = pg_get_querydef(query, pretty);

	AtEOXact_GUC(true, save_nestlevel);

	return newQuery;
}


/*
 * ExecuteCommand executes a SQL command.
 */
static void
ExecuteCommand(char *commandString)
{
	List	   *commandList = pg_parse_query(commandString);
	ListCell   *commandCell = NULL;

	foreach(commandCell, commandList)
	{
		RawStmt    *rawParseTree = lfirst(commandCell);

		PlannedStmt *plannedStmt = makeNode(PlannedStmt);

		plannedStmt->commandType = CMD_UTILITY;
		plannedStmt->utilityStmt = rawParseTree->stmt;

		ProcessUtility(plannedStmt, commandString, false, PROCESS_UTILITY_QUERY,
					   NULL, NULL, None_Receiver, NULL);
	}
}
