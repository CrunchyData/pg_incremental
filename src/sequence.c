#include "postgres.h"
#include "fmgr.h"
#include "funcapi.h"
#include "miscadmin.h"

#include "catalog/dependency.h"
#include "catalog/pg_authid.h"
#include "crunchy/incremental/pipeline.h"
#include "crunchy/incremental/sequence.h"
#include "executor/spi.h"
#include "storage/lmgr.h"
#include "storage/lock.h"
#include "utils/acl.h"
#include "utils/builtins.h"
#include "utils/lsyscache.h"
#include "utils/palloc.h"
#include "utils/rel.h"
#include "utils/snapmgr.h"


/*
 * SequenceNumberRange represents a range of sequence numbers that can
 * be safely processed.
 */
typedef struct SequenceNumberRange
{
	uint64		rangeStart;
	uint64		rangeEnd;
}			SequenceNumberRange;


static SequenceNumberRange * PopSequenceNumberRange(char *pipelineName, Oid sequenceId);
static SequenceNumberRange * GetSequenceNumberRange(char *pipelineName);


PG_FUNCTION_INFO_V1(incremental_sequence_range);


/*
 * InitializeSequencePipelineStats adds the initial sequence pipeline state.
 */
void
InitializeSequencePipelineState(char *pipelineName, Oid sequenceId)
{
	Oid			savedUserId = InvalidOid;
	int			savedSecurityContext = 0;

	/*
	 * Switch to superuser in case the current user does not have write
	 * privileges for the pipelines table.
	 */
	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
	SetUserIdAndSecContext(BOOTSTRAP_SUPERUSERID, SECURITY_LOCAL_USERID_CHANGE);

	char	   *query =
		"insert into incremental.sequence_pipelines "
		"(pipeline_name, sequence_name) "
		"values ($1, $2)";

	bool		readOnly = false;
	int			tupleCount = 0;
	int			argCount = 2;
	Oid			argTypes[] = {TEXTOID, OIDOID};
	Datum		argValues[] = {
		CStringGetTextDatum(pipelineName),
		ObjectIdGetDatum(sequenceId)
	};
	char	   *argNulls = "   ";

	SPI_connect();
	SPI_execute_with_args(query,
						  argCount,
						  argTypes,
						  argValues,
						  argNulls,
						  readOnly,
						  tupleCount);
	SPI_finish();

	SetUserIdAndSecContext(savedUserId, savedSecurityContext);
}


/*
 * incremental_sequence_range determines a safe range that can be processed
 * and stores progress as part of the transaction.
 */
void
ExecuteSequenceRangePipeline(char *pipelineName, char *command)
{
	PipelineDesc *pipelineDesc = ReadPipelineDesc(pipelineName);

	SequenceNumberRange *range =
		PopSequenceNumberRange(pipelineName, pipelineDesc->sourceRelationId);

	if (range->rangeStart > range->rangeEnd)
	{
		ereport(NOTICE, (errmsg("pipeline %s: no rows to process",
								pipelineName)));
		return;
	}

	ereport(NOTICE, (errmsg("pipeline %s: processing sequence values from "
							INT64_FORMAT " to " INT64_FORMAT,
							pipelineName, range->rangeStart, range->rangeEnd)));

	PushActiveSnapshot(GetTransactionSnapshot());

	bool		readOnly = false;
	int			tupleCount = 0;
	int			argCount = 2;
	Oid			argTypes[] = {INT8OID, INT8OID};
	Datum		argValues[] = {
		Int64GetDatum(range->rangeStart),
		Int64GetDatum(range->rangeEnd)
	};
	char	   *argNulls = "  ";

	SPI_connect();
	SPI_execute_with_args(command,
						  argCount,
						  argTypes,
						  argValues,
						  argNulls,
						  readOnly,
						  tupleCount);
	SPI_finish();

	PopActiveSnapshot();
}


/*
 * PopSequenceNumber range returns a range of sequence numbers that can
 * be safely processed by taking the last returned sequence number as the
 * end of the range, and waiting for all concurrent writers to finish.
 *
 * The start of the range is the end of the previous range + 1.
 *
 * Note: An assumptions is that writers will only insert sequence numbers
 * that were obtained after locking the table.
 */
static SequenceNumberRange *
PopSequenceNumberRange(char *pipelineName, Oid relationId)
{
	SequenceNumberRange *range = GetSequenceNumberRange(pipelineName);

	if (range->rangeStart <= range->rangeEnd)
	{
		LOCKTAG		tableLockTag;

		SET_LOCKTAG_RELATION(tableLockTag, MyDatabaseId, relationId);

		/*
		 * Wait for concurrent writers that may have seen sequence numbers <=
		 * the last-drawn sequence number.
		 */
		WaitForLockers(tableLockTag, ShareLock, true);

		/*
		 * We update the last-processed sequence number, which will commit or
		 * abort with the current (sub)transaction.
		 */
		UpdateLastProcessedSequenceNumber(pipelineName, range->rangeEnd);
	}

	return range;
}


/*
 * GetSequenceNumberRange reads the current state of the given sequence pipeline
 * and returns whether there are rows to process.
 */
static SequenceNumberRange *
GetSequenceNumberRange(char *pipelineName)
{
	SequenceNumberRange *range = (SequenceNumberRange *) palloc0(sizeof(SequenceNumberRange));

	Oid			savedUserId = InvalidOid;
	int			savedSecurityContext = 0;

	/*
	 * Switch to superuser in case the current user does not have write
	 * privileges for the pipelines table.
	 */
	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
	SetUserIdAndSecContext(BOOTSTRAP_SUPERUSERID, SECURITY_LOCAL_USERID_CHANGE);

	/*
	 * Get the last-drawn sequence number, which may be part of a write that
	 * has not committed yet. Also block other pipeline rollups.
	 */
	char	   *query =
		"select"
		" last_processed_sequence_number + 1,"
		" pg_catalog.pg_sequence_last_value(sequence_name) seq "
		"from incremental.sequence_pipelines "
		"where pipeline_name operator(pg_catalog.=) $1 "
		"for update";

	bool		readOnly = false;
	int			tupleCount = 0;
	int			argCount = 1;
	Oid			argTypes[] = {TEXTOID};
	Datum		argValues[] = {
		CStringGetTextDatum(pipelineName)
	};
	char	   *argNulls = " ";

	SPI_connect();
	SPI_execute_with_args(query,
						  argCount,
						  argTypes,
						  argValues,
						  argNulls,
						  readOnly,
						  tupleCount);

	if (SPI_processed <= 0)
		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT),
						errmsg("pipeline \"%s\" cannot be found",
							   pipelineName)));

	TupleDesc	rowDesc = SPI_tuptable->tupdesc;
	HeapTuple	row = SPI_tuptable->vals[0];

	/* read last_processed_sequence_number + 1 result */
	bool		rangeStartIsNull = false;
	Datum		rangeStartDatum = SPI_getbinval(row, rowDesc, 1, &rangeStartIsNull);

	if (!rangeStartIsNull)
		range->rangeStart = DatumGetInt64(rangeStartDatum);

	/* read pg_sequence_last_value result */
	bool		rangeEndIsNull = false;
	Datum		rangeEndDatum = SPI_getbinval(row, rowDesc, 2, &rangeEndIsNull);

	if (!rangeEndIsNull)
		range->rangeEnd = DatumGetInt64(rangeEndDatum);

	SPI_finish();

	SetUserIdAndSecContext(savedUserId, savedSecurityContext);

	/* return whether there are rows to process */
	return range;
}

/*
 * UpdateLastProcessedSequenceNumber updates the last_processed_sequence_number
 * in pipeline.pipelines to the given values.
 */
void
UpdateLastProcessedSequenceNumber(char *pipelineName, int64 lastSequenceNumber)
{
	Oid			savedUserId = InvalidOid;
	int			savedSecurityContext = 0;

	/*
	 * Switch to superuser in case the current user does not have write
	 * privileges for the pipleines table.
	 */
	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
	SetUserIdAndSecContext(BOOTSTRAP_SUPERUSERID, SECURITY_LOCAL_USERID_CHANGE);

	/*
	 * Get the last-drawn sequence number, which may be part of a write that
	 * has not committed yet. Also block other pipeline rollups.
	 */
	char	   *query =
		"update incremental.sequence_pipelines "
		"set last_processed_sequence_number = $2 "
		"where pipeline_name operator(pg_catalog.=) $1";

	bool		readOnly = false;
	int			tupleCount = 0;
	int			argCount = 2;
	Oid			argTypes[] = {TEXTOID, INT8OID};
	Datum		argValues[] = {
		CStringGetTextDatum(pipelineName),
		Int64GetDatum(lastSequenceNumber)
	};
	char	   *argNulls = "  ";

	SPI_connect();
	SPI_execute_with_args(query,
						  argCount,
						  argTypes,
						  argValues,
						  argNulls,
						  readOnly,
						  tupleCount);

	if (SPI_processed <= 0)
		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT),
						errmsg("pipeline \"%s\" cannot be found",
							   pipelineName)));

	SPI_finish();

	SetUserIdAndSecContext(savedUserId, savedSecurityContext);
}


/*
 * FindSequenceForRelation returns the Oid of a sequence belonging to the
 * given relation.
 *
 * We currently don't detect sequences that were manually added through
 * DEFAULT nextval(...).
 */
Oid
FindSequenceForRelation(Oid relationId)
{
	List	   *sequences = getOwnedSequences(relationId);

	if (list_length(sequences) == 0)
		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT),
						errmsg("relation \"%s\" does not have any sequences associated "
							   "with it",
							   get_rel_name(relationId))));

	if (list_length(sequences) > 1)
		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_OBJECT),
						errmsg("relation \"%s\" has multiple sequences associated "
							   "with it",
							   get_rel_name(relationId)),
						errhint("Specify the name of the sequence to use instead of "
								"the table name")));

	return linitial_oid(sequences);
}
