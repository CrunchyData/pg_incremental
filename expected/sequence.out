create extension pg_incremental cascade;
NOTICE:  installing required extension "pg_cron"
create schema sequence;
set search_path to sequence;
-- create a source table
create table events (
  event_id bigint generated always as identity,
  event_time timestamptz default now(),
  client_id bigint,
  path text,
  response_time double precision
);
-- BRIN indexes are highly effective in selecting new ranges
create index on events using brin (event_id);
-- generate some random inserts
insert into events (client_id, path, response_time)
select s, '/page-' || (s % 3), random() from generate_series(1,100) s;
-- create a summary table to pre-aggregate the number of events per day
create table events_agg (
  day timestamptz,
  event_count bigint,
  primary key (day)
);
select incremental.create_sequence_pipeline('event-aggregation',
  source_table_name := 'events',
  schedule := NULL,
  command := $$
  insert into events_agg
  select date_trunc('day', event_time), count(*)
  from events
  where event_id between $1 and $2
  group by 1
  on conflict (day) do update set event_count = events_agg.event_count + excluded.event_count;
  $$);
NOTICE:  pipeline event-aggregation: processing sequence values from 0 to 100
 create_sequence_pipeline 
--------------------------
 
(1 row)

select count(*) from events;
 count 
-------
   100
(1 row)

select sum(event_count) from events_agg;
 sum 
-----
 100
(1 row)

call incremental.execute_pipeline('event-aggregation');
NOTICE:  pipeline event-aggregation: no rows to process
insert into events (client_id, path, response_time)
select s, '/page-' || (s % 3), random() from generate_series(1,100) s;
call incremental.execute_pipeline('event-aggregation');
NOTICE:  pipeline event-aggregation: processing sequence values from 101 to 200
select count(*) from events;
 count 
-------
   200
(1 row)

select sum(event_count) from events_agg;
 sum 
-----
 200
(1 row)

create table events_json (id bigint generated by default as identity, payload jsonb);
/* periodically unpack the new JSON objects into the events table */
select incremental.create_sequence_pipeline('unpack-json', 'events_json', 
  schedule := NULL,
  command := $$
    insert into events (client_id, event_time, response_time)
    select '3', (payload->>'created_at')::timestamptz, random()
    from events_json
    where id between $1 and $2
  $$
);
NOTICE:  pipeline unpack-json: processing sequence values from 0 to 0
 create_sequence_pipeline 
--------------------------
 
(1 row)

insert into events_json (payload) values ('{"created_at":"2024-01-01 03:00:00"}');
call incremental.execute_pipeline('unpack-json');
NOTICE:  pipeline unpack-json: processing sequence values from 1 to 1
call incremental.execute_pipeline('event-aggregation');
NOTICE:  pipeline event-aggregation: processing sequence values from 201 to 201
select count(*) from events;
 count 
-------
   201
(1 row)

select sum(event_count) from events_agg;
 sum 
-----
 201
(1 row)

drop schema sequence cascade;
NOTICE:  drop cascades to 3 other objects
DETAIL:  drop cascades to table events
drop cascades to table events_agg
drop cascades to table events_json
drop extension pg_incremental;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to function incremental._drop_extension_trigger()
drop cascades to event trigger incremental_drop_extension_trigger
